components/script/dom/autofill.rs

```diff
@@ -0,0 +1,356 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use html5ever::local_name;
+use servo_atoms::Atom;
+use style::str::split_html_space_chars;
+
+use self::AutofillCategory::*;
+use crate::dom::bindings::import::module::Castable;
+use crate::dom::bindings::str::DOMString;
+use crate::dom::element::Element;
+use crate::dom::htmlformelement::FormControl;
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum AutofillCategory {
+    Off,
+    Automatic,
+    Normal,
+    Contact,
+    Credential,
+}
+
+const FIELD_CATEGORY_MAPPING: &[(&str, AutofillCategory)] = &[
+    ("additional-name", Normal),
+    ("address-level1", Normal),
+    ("address-level2", Normal),
+    ("address-level3", Normal),
+    ("address-level4", Normal),
+    ("address-line1", Normal),
+    ("address-line2", Normal),
+    ("address-line3", Normal),
+    ("bday", Normal),
+    ("bday-day", Normal),
+    ("bday-month", Normal),
+    ("bday-year", Normal),
+    ("cc-additional-name", Normal),
+    ("cc-csc", Normal),
+    ("cc-exp", Normal),
+    ("cc-exp-month", Normal),
+    ("cc-exp-year", Normal),
+    ("cc-family-name", Normal),
+    ("cc-given-name", Normal),
+    ("cc-name", Normal),
+    ("cc-number", Normal),
+    ("cc-type", Normal),
+    ("country", Normal),
+    ("country-name", Normal),
+    ("current-password", Normal),
+    ("device-eid", Normal),
+    ("device-imei", Normal),
+    ("email", Contact),
+    ("family-name", Normal),
+    ("given-name", Normal),
+    ("honorific-prefix", Normal),
+    ("honorific-suffix", Normal),
+    ("impp", Contact),
+    ("language", Normal),
+    ("name", Normal),
+    ("new-password", Normal),
+    ("nickname", Normal),
+    ("off", Off),
+    ("on", Automatic),
+    ("one-time-code", Normal),
+    ("organization", Normal),
+    ("organization-title", Normal),
+    ("photo", Normal),
+    ("postal-code", Normal),
+    ("sex", Normal),
+    ("street-address", Normal),
+    ("tel", Contact),
+    ("tel-area-code", Contact),
+    ("tel-country-code", Contact),
+    ("tel-extension", Contact),
+    ("tel-local", Contact),
+    ("tel-local-prefix", Contact),
+    ("tel-local-suffix", Contact),
+    ("tel-national", Contact),
+    ("transaction-amount", Normal),
+    ("transaction-currency", Normal),
+    ("url", Normal),
+    ("username", Normal),
+    ("webauthn", Credential),
+];
+
+#[derive(Debug, PartialEq)]
+pub enum NonAutofillCredentialType {
+    WebAuthn,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum AutofillMantle {
+    Anchor,
+    Expectation,
+}
+
+#[allow(unused)]
+#[derive(Debug)]
+pub struct AutofillData {
```

Comment:
I think we should store this on the elements themselves via a `DomRefCell<Option>>`, and re-run the algorithm when necessary via the [`attribute_mutated`](https://github.com/servo/servo/blob/b243457ccc6cd5a2dab58d9c9ff8b6fee1db6a20/components/script/dom/htmlinputelement.rs#L2258) hook.